### ISSUE TITLE
Namespace required when extension function has same name as return type

### ISSUE THREAD

**Issue by lpavel** [dotnet/roslyn#33661]
**Version Used**: VS 15.9.4

I got across this behavior today which seemed surprising.

For the following code:
```
namespace Authentication
{
    public static class RestrictedClaimTypes
    {
        public const string AppIdAcr = "appidacr";
    }

    public enum AppAuthType
    {
        None = -1,
        PublicClient = 0,
        ClientIdAndSecret = 1,
        ClientCertificate = 2
    }

    public static AppAuthType AppAuthType(this ClaimsPrincipal user)
    {
        string appIdAcr = user.FindFirst(RestrictedClaimTypes.AppIdAcr)?.Value;
        if (!Enum.TryParse<AppAuthType>(appIdAcr, out var authType))
        {
            return AppAuthType.None;
        }
        return authType;
    }
}
```

This fails to build with the following error: 
error CS0119: 'ClaimsPrincipalExtensions.AppAuthType(ClaimsPrincipal)' is a method, which is not valid in the given context 

However, adding the namespace of the AppAuthType fixes it: 
`return AppAuthType.None;` -> `return Authentication.AppAuthType.None;`

Basically, it looks like without the `Authentication.`, it is trying to match AppAuthType with the extension function. I find it a bit confusing to why it does that and was expecting it to work without the namespace. Is this behavior expected?

**Comment by CyrusNajmabadi** [comment]
>  it is trying to match AppAuthType with the extension function. I find it a bit confusing to why it does that and was expecting it to work without the namespace. Is this behavior expected?

This is expected.  You wrote `return AppAuthType...`.  So the compiler goes to lookup AppAuthType.  It walks upwards in scope and immediately finds the containing `AppAuthType` method.  So it has bound to that.  You then `dot` off of it.  That's illegal to do off a method.  So it fails.

The language has no concept of `you later did something illegal, so let me walk backwards and then try this again and see if i find something that would make it legal`.  It just consistently says "if you're in the location and you're in a value-or-type spot, and you write this name, lookup will produce this value."

**Comment by Joe4evr** [comment]
To add on to Cyrus' explanation, this is why [the naming guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-type-members#names-of-methods) want your methods to be named with a verb or verb-phrase. In this case I would suggest `GetAppAuthType` to resolve the ambiguity.

**Comment by lpavel** [comment]
Thank you very much for the explanation and the suggestion.
